src/Blocks/BlockButton.h

```diff
@@ -133,12 +123,110 @@ class cBlockButtonHandler :
 	}
 
 	/** Extracts the ON bit from metadata and returns if true if it is set */
-	static bool IsButtonOn(NIBBLETYPE a_BlockMeta)
+	static bool IsButtonOn(NIBBLETYPE a_Meta)
 	{
-		return ((a_BlockMeta & 0x8) == 0x8);
+		return (a_Meta & 0x08) == 0x08;
+	}
+
+	/** Event handler for an arrow striking a block.
+	Performs appropriate handling if the arrow intersected a wooden button. */
+	static void OnArrowHit(cWorld & a_World, const Vector3i a_Position, const eBlockFace a_HitFace)
+	{
+		BLOCKTYPE Type;
+		NIBBLETYPE Meta;
+		const auto Pos = AddFaceDirection(a_Position, a_HitFace);
+
+		if (
+			!a_World.GetBlockTypeMeta(Pos, Type, Meta) ||
+			(Type != E_BLOCK_WOODEN_BUTTON) || IsButtonOn(Meta) ||
+			!IsButtonPressedByArrow(a_World, Pos, Type, Meta)
+		)
+		{
+			// Bail if we're not specifically a wooden button, or it's already on
+			// or if the arrow didn't intersect. It is very important that nothing is
+			// done if the button is depressed, since the release task will already be queued
+			return;
+		}
+
+		a_World.SetBlockMeta(Pos, Meta | 0x08, false);
+		a_World.WakeUpSimulators(Pos);
+
+		// sound name is ok to be wood, because only wood gets triggered by arrow
+		a_World.GetBroadcastManager().BroadcastSoundEffect("block.wood_button.click_on", Pos, 0.5f, 0.6f);
+
+		// Queue a button reset
+		QueueButtonRelease(a_World, Pos, Type);
+	}
+
+private:
+
+	/** Schedules a recurring event at appropriate intervals to release a button at a given position.
+	The given block type is checked when the task is executed to ensure the position still contains a button. */
+	static void QueueButtonRelease(cWorld & a_World, const Vector3i a_Position, const BLOCKTYPE a_BlockType)
+	{
+		const auto TickDelay = (a_BlockType == E_BLOCK_STONE_BUTTON) ? 20 : 30;
+		a_World.ScheduleTask(
+			TickDelay,
+			[a_Position, a_BlockType](cWorld & a_World)
+			{
+				BLOCKTYPE Type;
+				NIBBLETYPE Meta;
+
+				if (
+					!a_World.GetBlockTypeMeta(a_Position, Type, Meta) ||
+					(Type != a_BlockType) ||
```

Comment:
So when the button has been mined away this scheduled task will run continually?

Reply:
good point